torch/_dynamo/variables/builder.py

```diff
@@ -1500,6 +1500,7 @@ def _clone_input(value):
         torch.sym_int,
         getattr,
         operator.getitem,
+        operator.eq,
```

Comment:
Why did we need this again?

Reply:
Because the eq comparison in `assert_non_empty_tensor_output`:
```
[2024-02-21 12:55:29,233] [0/0] torch._dynamo.output_graph.__trace_call: [DEBUG] TRACE FX call eq from /home/guilhermeleobas/git/pytorch/torch/_functorch/eager_transforms.py:830 in assert_non_empty_tensor_output (assert_non_empty_tensor_output) (inline depth: 4)
[2024-02-21 12:55:29,233] [0/0] torch._dynamo.output_graph.__trace_call: [DEBUG]     if output == [None] or len(output) < 1:
[2024-02-21 12:55:29,233] [0/0] torch._dynamo.output_graph.__trace_call: [DEBUG]        ^^^^^^^^^^^^^^^^
```

This falls into a generic list comparison, which iterates over each element and check if they're equal:
https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/lists.py#L165-L179

As for reference, below are the items in both `left` and `right`:
```python
(Pdb++) left
ListVariable()
(Pdb++) left.items
[TensorVariable()]
(Pdb++) right
ListVariable()
(Pdb++) right.items
[ConstantVariable(NoneType: None)]
```

Reply:
I tested the three tests cases you said, and two of them work.

```
import torch

@torch.compile(dynamic=True)
def foo():
    a = torch.randn([])
    b = torch.randn([])
    return a == b


@torch.compile(dynamic=True)
def bar():
    a = torch.randn([])
    # this is not actually valid
    b = a.size(0)
    return a == b


@torch.compile(dynamic=True)
def baz():
    a = torch.randn(3, 4)
    return a.size(0) == a.size(1)


# should return a boolean tensor
print(foo())  # tensor(False)

# either a boolean tensor or a graph break
# print(bar())  # test case is not valid

# either a symbolic bool or a graph break
print(baz())  # False
```

Reply:
Do you know where `Tensor.__eq__` is? My questions here are:
1) When does operator.eq return an int
2) it's likely that we should not emit a graph with operator.eq in it, because it is not a Tensor operation, so adding operator.eq to this list may not be the correct approach.

Reply:
I figured it out:
0. `Tensor.__eq__` runs torch.eq, but if the types are not what we expect, it returns NotImplemented
1. operator.eq returns an int when torch.eq returns NotImplemented
2. Since operator.eq in this context is not a Tensor operation, it shouldn't go into the graph. Dynamo should just return a ConstantVariable(...) when it sees it. We shouldn't handle it inside this function the comment is pointing to, because this function is responsible for putting operators into the graph.

Reply:
Btw, an easy workaround is to change the original condition to `len(output) == 1 and output[0] is None`

Reply:
I've changed the code in `assert_non_empty_tensor_output` to not compare lists and including the `operator.eq` in the list won't be needed anymore. Could you review again?

> it's likely that we should not emit a graph with operator.eq in it, because it is not a Tensor operation, so adding operator.eq to this list may not be the correct approach.

Yes, you're right.